lib/client/ca_export.go

```diff
@@ -0,0 +1,289 @@
+
+	// this means to export TLS authority
+	switch req.AuthType {
+	// "tls" is supported for backwards compatibility.
+	// "tls-host" and "tls-user" were added later to allow export of the user
+	// TLS CA.
+	case "tls", "tls-host":
+		req := exportTLSAuthorityRequest{
+			AuthType:          types.HostCA,
+			UnpackPEM:         false,
+			ExportPrivateKeys: req.ExportPrivateKeys,
+		}
+		return exportTLSAuthority(ctx, client, req)
+	case "tls-user":
+		req := exportTLSAuthorityRequest{
+			AuthType:          types.UserCA,
+			UnpackPEM:         false,
+			ExportPrivateKeys: req.ExportPrivateKeys,
+		}
+		return exportTLSAuthority(ctx, client, req)
+	case "db":
+		req := exportTLSAuthorityRequest{
+			AuthType:          types.DatabaseCA,
+			UnpackPEM:         false,
+			ExportPrivateKeys: req.ExportPrivateKeys,
+		}
+		return exportTLSAuthority(ctx, client, req)
+	case "tls-user-der", "windows":
+		req := exportTLSAuthorityRequest{
+			AuthType:          types.UserCA,
+			UnpackPEM:         true,
+			ExportPrivateKeys: req.ExportPrivateKeys,
+		}
+		return exportTLSAuthority(ctx, client, req)
+	}
+
+	// If no AuthType is given, export both HostCA and UserCA.
+	if req.AuthType == "" {
+		typesToExport = []types.CertAuthType{types.HostCA, types.UserCA}
+	} else {
+		authType := types.CertAuthType(req.AuthType)
+		if err := authType.Check(); err != nil {
+			return "", trace.Wrap(err)
+		}
+		typesToExport = []types.CertAuthType{authType}
+	}
+	localAuthName, err := client.GetDomainName(ctx)
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+
+	// fetch authorities via auth API (and only take local CAs, ignoring
+	// trusted ones)
+	var authorities []types.CertAuthority
+	for _, at := range typesToExport {
+		cas, err := client.GetCertAuthorities(ctx, at, req.ExportPrivateKeys)
+		if err != nil {
+			return "", trace.Wrap(err)
+		}
+		for _, ca := range cas {
+			if ca.GetClusterName() == localAuthName {
+				authorities = append(authorities, ca)
+			}
+		}
+	}
+
+	ret := strings.Builder{}
+	for _, ca := range authorities {
+		if req.ExportPrivateKeys {
+			for _, key := range ca.GetActiveKeys().SSH {
+				fingerprint, err := sshutils.PrivateKeyFingerprint(key.PrivateKey)
+				if err != nil {
+					return "", trace.Wrap(err)
+				}
+				if req.ExportAuthorityFingerprint != "" && fingerprint != req.ExportAuthorityFingerprint {
+					continue
+				}
+				ret.WriteString(string(key.PrivateKey))
+				ret.WriteString("\n")
+			}
+			continue
+		}
+
+		for _, key := range ca.GetTrustedSSHKeyPairs() {
+			fingerprint, err := sshutils.AuthorizedKeyFingerprint(key.PublicKey)
+			if err != nil {
+				return "", trace.Wrap(err)
+			}
+			if req.ExportAuthorityFingerprint != "" && fingerprint != req.ExportAuthorityFingerprint {
+				continue
+			}
+
+			// export certificates in the old 1.0 format where host and user
+			// certificate authorities were exported in the known_hosts format.
+			if req.CompatVersion == "1.0" {
```

Comment:
Should we check bad compat versions earlier and error? (Or make it a bool, per previous comment?)

Reply:
Turning this into a bool